### PR TITLE
Allows to purchase tickets

### LINKED ISSUE #1: Implement Mercado Pago API calls

### PR DESCRIPTION
**Pull request by joelalejandro**
This PR adds the initial support to purchase a ticket.

### Preparing locally
- Run `yarn` to install the new dependencies
- Export an environment variable called `DATABASE_URL`, with a connection string pointing to a local instance of PostgreSQL (>10.x).
- Run `yarn migrate:latest && yarn seed:run` to create all the tables and load the minimum data required.
- Export environment variables to configure Mercado Pago's gateway (see values [here](https://docs.google.com/spreadsheets/d/1BWFpUSc2p7i9nEW8ZNYDuyInQmpccl_xTiic7D4sFG4/edit#gid=1247615833)).
- Run `yarn start:dev`

### Order of API calls
In order for a purchase to be completed, we need to:

1. Save a customer for each ticket owner.
2. Book tickets for each customer.
3. Create a purchase with a paying customer, selecting all the tickets included.

#### Step 1: Save a customer for each ticket owner

A customer is a potential owner of a ticket.

```json
// POST /api/customers
{
  "data": {
    "type": "customer",
    "attributes": {
      "fullName": "Happy Webconfer",
      "identificationType": "DNI",
      "identificationNumber": "12345678"
    }
  }
}
```

#### Step 2: Book tickets for each customer
Create a ticket for the customer. This will select the first ticket on status `forSale` of type `attendee` (remember: all tickets are pre-created in the database, we **never** create tickets so we can keep a strict control of the amount of tickets available).

```json
// POST /api/tickets
{
  "data": {
    "type": "ticket",
    "attributes": {},
    "relationships": {
      "customer": {
        "data": {
          "id": "e712aeeb-a2c7-4256-9327-0233933724b8",
          "type": "customer"
        }
      }
    }
  }
}
```

#### Step 3: Create the purchase
The Purchase API triggers a payment creation with Mercado Pago. A list of ticket IDs must be passed.

```json
{
  "data": {
    "type": "purchase",
    "attributes": {},
    "relationships": {
      "ticket": {"data": [{ "id": "09aec77d-7cc2-4c21-908b-8880f034c191", "type": "ticket" }]}
    }
  }
}
```

This PR also enables:

- Sending an e-mail to the customer, with a receipt + the tickets.
- Generating QR codes for the tickets.
- Validating and rolling back any changes in case of failure.

Resolves #1 (kinda).

Full documentation will live [here](https://github.com/WebConfTech/checkout-api/wiki).

**Pending**
- Securing the API with a private app token, so no other than the Checkout Frontend can actually call this API.

**Pending for further releases**
- Adding an IPN (Instant Payment Notification) endpoint to update our data when the customer actually pays.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (homer0) [inline .vscode/launch.json]: `.gitignore`

File: package.json [hunk 2]
```diff
@@ -18,15 +23,20 @@
...
-    "@ebryn/jsonapi-ts": "ebryn/jsonapi-ts",
+    "@joelalejandro/jsonapi-ts": "joelalejandro/jsonapi-ts",
     "@koa/cors": "^2.2.3",
+    "knex": "^0.16.3",
     "koa": "^2.7.0",
-    "todo-pago": "TodoPago/SDK-Node.js"
+    "logepi": "^1.0.5",
+    "mercadopago": "^1.0.22",
+    "pg": "^7.9.0",
+    "ts-node-dev": "^1.0.0-pre.32",
```
Comment (homer0) [inline package.json:40]: Why did you move it to a prod dep?

File: src/index.ts [hunk 3]
```diff
@@ -1,28 +1,66 @@
...
 const api = new Koa();
+let connection = process.env.DATABASE_URL;
+
+if (process.env.NODE_ENV !== "development") {
+  // Enforce SSL for DB connections.
+  connection += "?ssl=true";
+
+  // Enforce SSL for HTTP requests.
+  api.use(async (ctx: Koa.Context, next: () => Promise<void>) => {
+    if (ctx.headers["x-forwarded-proto"] !== "https") {
```
Comment (homer0) [inline src/index.ts]: Why use the header instead/with `.secure`?

File: src/index.ts [hunk 4]
```diff
@@ -1,28 +1,66 @@
...
       ]
     })
-  )
-);
+  );
+
+api
+  .use(cors())
+  .use(checkout())
+  .use(logs());
```
Comment (homer0) [inline src/index.ts:71]: Shouldn't this be only on dev?

**Review by asermax — APPROVED** [review]
Overall looks okay to me; I have some doubts regarding how we're going to be using this API (how we  obtain the final price for a ticket + credit card interest for example) but I guess  those things are gonna start popping up as we go forward with the frontend implementation.

**Review by homer0 — CHANGES_REQUESTED** [review]

File: .vscode/launch.json (hunk 1 above)
Comment (joelalejandro) [reply .vscode/launch.json]: Why? It's nice to have a pre-opinionated debug configuration.

File: package.json (hunk 2 above)
Comment (joelalejandro) [reply package.json:40]: Ah, this is a leftover. Just eagerly testing if the app would run.

File: src/index.ts (hunk 3 above)
Comment (joelalejandro) [reply src/index.ts]: Didn't know about .secure() xD

File: src/index.ts (hunk 4 above)
Comment (joelalejandro) [reply src/index.ts:71]: Nope. We want logs on production.

**Comment by joelalejandro** [comment]
@asermax The actual call order and integration with the front-end is being described on this repo's wiki.

File: .vscode/launch.json (hunk 1 above)
Comment (homer0) [reply .vscode/launch.json]: Ok, I'll probably never code on this project... but the usual reasons: not everyone uses the same IDE/editor. This is not project related, is personal preference.

Comment (joelalejandro) [reply .vscode/launch.json]: I'll just remove it. :P

**Review by homer0 — APPROVED** [review]

File: src/resources/purchase/processor.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    // 2: Create preference.
+    const preference = await this.postPreference({ tickets, quantitiesByType });
+
+    // 3: Create the purchase.
+    purchase.id = uuid.v4();
+    purchase.attributes = {
+      dateCreated: new Date().toJSON(),
+      status: "unpaid",
+      amountBilled: totalPrice,
+      externalId: preference.id
```
Comment (asermax) [inline src/resources/purchase/processor.ts:33]: To generate the payment URL I should use `https://www.mercadopago.com/mla/checkout/start?pref_id=<preference id>` right?

Comment (joelalejandro) [reply src/resources/purchase/processor.ts:33]: Yes!

**Comment by asermax** [comment]
I found an issue when trying to click on a link that says "Volver a WebConf" on the payment site.

image

When I click there, I get redirected to the following url: https://checkout.webconf.tech/webhooks/purchase-failure?collection_id=null&collection_status=null&preference_id=412901074-7051a0b6-32c8-4120-9140-1a5bb15b3f3c&external_reference=9ff292c5-bf98-46c9-a376-67dcafe252bb&payment_type=credit_card&merchant_order_id=null

The url responds with an `Internal Server Error`; here's the error I see on Heroku's logs:
```
2019-03-28T15:36:48.555401+00:00 heroku[router]: at=info method=GET path="/webhooks/purchase-failure?collection_id=null&collection_status=null&preference_id=412901074-7051a0b6-32c8-4120-9140-1a5bb15b3f3c&external_reference=9ff292c5-bf98-46c9-a376-67dcafe252bb&payment_type=credit_card&merchant_order_id=null" host=checkout.webconf.tech request_id=3ea4f8aa-7c24-431d-ab15-a579b272eced fwd="181.171.104.246" dyno=web.1 connect=0ms service=1ms status=500 bytes=176 protocol=https
2019-03-28T15:36:48.554090+00:00 app[web.1]: 
2019-03-28T15:36:48.554146+00:00 app[web.1]:   TypeError: Cannot read property 'toLowerCase' of undefined
2019-03-28T15:36:48.554149+00:00 app[web.1]:       at Object.singularize (/app/node_modules/pluralize/pluralize.js:133:24)
2019-03-28T15:36:48.554151+00:00 app[web.1]:       at jsonApiKoa (/app/node_modules/@joelalejandro/jsonapi-ts/dist/middlewares/json-api-koa.js:26:59)
2019-03-28T15:36:48.554153+00:00 app[web.1]:       at process._tickCallback (internal/process/next_tick.js:68:7)
```
